File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/CallKitManager.m
//
//  CallKit.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 26.05.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "CallKitManager.h"
#import "Log.h"
#import "CallKitInfo.h"

static const NSInteger DefaultMaximumCallsPerCallGroup = 1;
static const NSInteger DefaultMaximumCallGroups = 1;

@interface CallKitManager ()

@property (strong, nonatomic) CXProvider *provider;
@property (strong, nonatomic) CXCallController *callController;
@property (strong, nonatomic) CallKitInfo *call;
@property (strong, nonatomic) QBRTCAudioSession  *qbAudioSession;
@property (assign, nonatomic) BOOL reportEndCall;

@end

@interface CallKitManager (Provider) <CXProviderDelegate>
@end

@implementation CallKitManager

+ (CXProviderConfiguration *)configuration {
    CXProviderConfiguration *config = [[CXProviderConfiguration alloc] init];
    config.supportsVideo = YES;
    config.maximumCallsPerCallGroup = DefaultMaximumCallsPerCallGroup;
    config.maximumCallGroups = DefaultMaximumCallGroups;
    config.supportedHandleTypes = [NSSet setWithObjects:@(CXHandleTypePhoneNumber), nil];
    config.iconTemplateImageData = UIImagePNGRepresentation([UIImage imageNamed:@"qb-logo"]);
    config.ringtoneSound = @"ringtone.wav";
    return config;
}

- (NSUUID *)callUUID {
    return self.call.uuid;
}

// MARK: - Initialization
- (instancetype)init {
    self = [super init];
    if (self != nil) {
        CXProviderConfiguration *configuration = [CallKitManager configuration];
        self.provider = [[CXProvider alloc] initWithConfiguration:configuration];
        [self.provider setDelegate:self queue:nil];
        self.callController = [[CXCallController alloc] initWithQueue:dispatch_get_main_queue()];
        self.qbAudioSession = [QBRTCAudioSession instance];
        self.reportEndCall = NO;
    }
    return self;
}

// MARK: - Actions

- (void)reportIncomingCall:(NSString *)sessionId
                     title:(NSString *)title
                  hasVideo:(BOOL)hasVideo
                     state:(IncommingCallState)state
                completion:(void (^)(void))completion {
    CXCallUpdate *update = [self callUpdateWithTitle:title
                                            hasVideo:hasVideo];
    CallKitInfo *call = [[CallKitInfo alloc] initWithSessionId:sessionId hasVideo:hasVideo];
    self.call = call;
    [self.provider reportNewIncomingCallWithUUID:call.uuid
                                          update:update
                                      completion:^(NSError * _Nullable error) {
        if (completion != nil) {
            completion();
        }
        if (error) {
            Log(@"[%@] Error: %@",  NSStringFromClass(CallKitManager.class), error);
            return;
        }
        CXCallEndedReason reason;
        switch (state) {
            case IncommingCallStateValid:
                self.qbAudioSession.useManualAudio = YES;
                return;
            case IncommingCallStateMissed: reason = CXCallEndedReasonRemoteEnded;
                break;
            case IncommingCallStateInvalid: reason = CXCallEndedReasonUnanswered;
                break;
        }
        self.call = nil;
        [self.provider reportCallWithUUID:call.uuid
                              endedAtDate:NSDate.date
                                   reason:reason];
    }];
}

- (void)reportOutgoingCall:(NSString *)sessionId
                     title:(NSString *)title
                  hasVideo:(BOOL)hasVideo
                completion:(void (^)(void))completion {
    CallKitInfo *call = [[CallKitInfo alloc] initWithSessionId:sessionId hasVideo:hasVideo];
    self.call = call;
    CXHandle *handle = [self handleWithText:title];
    
    CXStartCallAction *action = [[CXStartCallAction alloc] initWithCallUUID:call.uuid
                                                                     handle:handle];
    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
    
    [self.callController requestTransaction:transaction completion:^(NSError *error) {
        if (error != nil) {
            Log(@"[%@] Error: %@",  NSStringFromClass(CallKitManager.class), error);
        }
        if (completion) { completion(); }
    }];
}

- (void)reportEndCall:(NSString *)sessionId {
    if (![sessionId isEqualToString:self.call.sessionId]) {
        return;
    }
    
    self.reportEndCall = YES;
    
    CXEndCallAction *action = [[CXEndCallAction alloc] initWithCallUUID:self.call.uuid];
    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
    
    [self.callController requestTransaction:transaction completion:^(NSError *error) {
        if (error == nil) {
            return;
        }
        Log(@"[%@] Error: %@",  NSStringFromClass(CallKitManager.class), error);
    }];
}

- (void)reportEndCall:(NSString *)sessionId reason:(CXCallEndedReason)reason {
    if (![sessionId isEqualToString:self.call.sessionId]) {
        return;
    }
    
    [self.provider reportCallWithUUID:self.call.uuid
                          endedAtDate:NSDate.date
                               reason:reason];
    [self closeCallWithSessionID:sessionId];
}

- (void)reportAcceptCall:(NSString *)sessionId {
    NSUUID *callUUID = [[NSUUID alloc] initWithUUIDString:sessionId];
    NSArray *actions = [self.provider pendingCallActionsOfClass:CXAnswerCallAction.class
                                                   withCallUUID:callUUID];
    for (CXAction *action in actions) {
        CXAnswerCallAction *answer = (CXAnswerCallAction *)action;
        [answer fulfillWithDateConnected:NSDate.date];
    }
}

- (void)reportUpdateCall:(NSString *)sessionId
                   title:(NSString *)title {
    if (!title.length) {
        return;
    }
    NSUUID *callUUID = [[NSUUID alloc] initWithUUIDString:sessionId];
    CXCallUpdate *update = [[CXCallUpdate alloc] init];
    update.remoteHandle = [self handleWithText:title];
    update.localizedCallerName = title;
    
    [self.provider reportCallWithUUID:callUUID updated:update];
}

- (void)muteAudio:(BOOL)mute call:(NSString *)sessionId {
    if (![sessionId isEqualToString:self.call.sessionId]) {
        return;
    }
    
    CXSetMutedCallAction *action = [[CXSetMutedCallAction alloc] initWithCallUUID:self.call.uuid muted:mute];
    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
    [self.callController requestTransaction:transaction completion:^(NSError *error) {
        if (error != nil) {
            Log(@"[%@] Error: %@",  NSStringFromClass(CallKitManager.class), error);
        }
    }];
}

- (void)closeCallWithSessionID:(NSString *)sessionID {
    self.qbAudioSession.audioEnabled = NO;
    self.qbAudioSession.useManualAudio = NO;
    if (self.reportEndCall) {
        if ([self.delegate respondsToSelector:@selector(callKit:didEndCall:)]) {
            [self.delegate callKit:self didEndCall:self.call.sessionId];
        }
    } else {
        if ([self.delegate respondsToSelector:@selector(callKit:didTapRedject:)]) {
            [self.delegate callKit:self didTapRedject:self.call.sessionId];
        }
    }
    
    self.reportEndCall = NO;
    self.call = nil;
}

//MARK: - Internal
- (CXCallUpdate *)callUpdateWithTitle:(NSString *)title
                             hasVideo:(BOOL)hasVideo {
    CXCallUpdate *update = [[CXCallUpdate alloc] init];
    update.localizedCallerName = title;
    update.supportsHolding = NO;
    update.supportsGrouping = NO;
    update.supportsUngrouping = NO;
    update.supportsDTMF = NO;
    update.hasVideo = hasVideo;
    
    return update;
}

- (CXHandle *)handleWithText:(NSString *)text {
    return [[CXHandle alloc] initWithType:CXHandleTypeGeneric value:text];
}

- (void)updateAudioSessionConfiguration:(BOOL)hasVideo {
    
    QBRTCAudioSessionConfiguration *configuration = [[QBRTCAudioSessionConfiguration alloc] init];
    configuration.categoryOptions |= AVAudioSessionCategoryOptionDuckOthers;
    
    // adding blutetooth support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
    
    // adding airplay support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
    
    if (hasVideo) {
        configuration.mode = AVAudioSessionModeVideoChat;
    }
    [self.qbAudioSession setConfiguration:configuration];
}

@end

@implementation CallKitManager (Provider)

- (void)providerDidReset:(nonnull CXProvider *)provider {
}

- (void)provider:(CXProvider *)provider performStartCallAction:(CXStartCallAction *)action {
    if ([action.callUUID isEqual:self.callUUID] == NO) {
        [action fail];
        return;
    }
    
    [self updateAudioSessionConfiguration:self.call.hasVideo];
    
    [action fulfill];
}

- (void)provider:(CXProvider *)provider performAnswerCallAction:(CXAnswerCallAction *)action {
    if ([action.callUUID isEqual:self.callUUID] == NO) {
        [action fail];
        return;
    }
    
    [self updateAudioSessionConfiguration:self.call.hasVideo];
    
    if ([self.delegate respondsToSelector:@selector(callKit:didTapAnswer:)]) {
        [self.delegate callKit:self didTapAnswer:self.call.sessionId];
    }
}

- (void)provider:(CXProvider *)provider performEndCallAction:(CXEndCallAction *)action {
    if ([action.callUUID isEqual:self.callUUID] == NO && self.call) {
        [action fail];
        return;
    }
    
    [self closeCallWithSessionID:self.call.sessionId];
    
    [action fulfill];
}

- (void)provider:(CXProvider *)provider didActivateAudioSession:(AVAudioSession *)audioSession {
    [self.qbAudioSession audioSessionDidActivate:audioSession];
    self.qbAudioSession.audioEnabled = YES;
}

- (void)provider:(CXProvider *)provider didDeactivateAudioSession:(AVAudioSession *)audioSession {
    [self.qbAudioSession audioSessionDidDeactivate:audioSession];
}

- (void)provider:(CXProvider *)__unused provider performSetMutedCallAction:(CXSetMutedCallAction *)action {
    if ([action.callUUID isEqual:self.callUUID] == NO) {
        [action fail];
        return;
    }
    if ([self.actionDelegate respondsToSelector:@selector(callKit:didTapMute:)]) {
        [self.actionDelegate callKit:self didTapMute:action.isMuted];
    }
    [action fulfill];
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/CallKitManager.swift
//
//  CallKitManager.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 30.03.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC
import Quickblox
import CallKit

typealias ReportCallCompletion = (() -> Void)

struct CallKitConstant {
    static let defaultMaximumCallsPerCallGroup: Int = 1
    static let defaultMaximumCallGroups: Int = 1
}

enum IncommingCallState: UInt {
    case valid = 0
    case missed
    /// Some call data wrong or absent
    case invalid
}

protocol CallKitManagerDelegate: AnyObject {
    func callKit(_ callKit: CallKitManager, didTapAnswer sessionId: String)
    func callKit(_ callKit: CallKitManager, didTapRedject sessionId: String)
    
    /// external ending using "reportEndCall" methods
    func callKit(_ callKit: CallKitManager, didEndCall sessionId: String)
}

protocol CallKitManagerActionDelegate: AnyObject {
    func callKit(_ callKit: CallKitManager, didTapMute isMuted: Bool)
}

final class CallKitManager: NSObject {
    
    //MARK: - Properties
    weak var delegate: CallKitManagerDelegate?
    weak var actionDelegate: CallKitManagerActionDelegate?
    private var provider: CXProvider?
    private let qbAudioSession = QBRTCAudioSession.instance()
    private let callController = CXCallController(queue: DispatchQueue.main)
    private var call: CallKitInfo?
    private var reportEndCall: Bool = false
    
    static let providerConfiguration: CXProviderConfiguration = {
        let config = CXProviderConfiguration()
        config.supportsVideo = true
        config.maximumCallsPerCallGroup = CallKitConstant.defaultMaximumCallsPerCallGroup
        config.maximumCallGroups = CallKitConstant.defaultMaximumCallGroups
        config.supportedHandleTypes = [.phoneNumber]
        if let image = UIImage(named: "qb-logo") {
            config.iconTemplateImageData = image.pngData()
        }
        config.ringtoneSound = "ringtone.wav"
        return config
    }()
    
    func callUUID() -> UUID? {
        return call?.uuid
    }
    
    // MARK: - Life Cycle
    override init() {
        provider = CXProvider(configuration: type(of: self).providerConfiguration)
        super.init()
        if let provider = provider {
            provider.setDelegate(self, queue: nil)
        }
    }
    
    //MARK: - Public Methods
    func reportIncomingCall(sessionId: String,
                            title: String,
                            hasVideo: Bool,
                            state: IncommingCallState,
                            completion: ReportCallCompletion? = nil) {
        
        guard let provider = provider else {
            return
        }
        
        let update = callUpdate(withTitle: title, hasVideo: hasVideo)
        let call = CallKitInfo(sessionID: sessionId, hasVideo: hasVideo)
        self.call = call
        
        provider.reportNewIncomingCall(with: call.uuid, update: update) { [weak self] error in
            defer { completion?() }
            if let error = error {
                debugPrint("\(#function) Error: \(error)")
                return
            }
            switch state {
            case .valid:
                self?.qbAudioSession.useManualAudio = true
                return
            case .missed: provider.reportCall(with: call.uuid, endedAt: Date(), reason: .remoteEnded)
            case .invalid: provider.reportCall(with: call.uuid, endedAt: Date(), reason: .unanswered)
            }
            self?.call = nil
        }
    }
    
    func reportOutgoingCall(sessionId: String,
                            title: String,
                            hasVideo: Bool,
                            completion: ReportCallCompletion? = nil) {
        
        call = CallKitInfo(sessionID: sessionId, hasVideo: hasVideo)
        
        guard let call = call else {
            return
        }
        
        let action = CXStartCallAction(call: call.uuid, handle: handle(withText: title))
        let transaction = CXTransaction(action: action)
        
        callController.request(transaction) { error in
            defer { completion?() }
            if let error = error {
                debugPrint("\(#function) Error: \(error)")
            }
        }
    }
    
    func reportEndCall(sessionId: String) {
        guard let call = call,
              sessionId == call.sessionID else {
            return
        }
        
        reportEndCall = true
        let  action = CXEndCallAction(call: call.uuid)
        let transaction = CXTransaction(action: action)
        callController.request(transaction) { error in
            guard let error = error else {
                return
            }
            debugPrint("\(#function) Error: \(error)")
        }
    }
    
    func reportEndCall(sessionId: String, reason: CXCallEndedReason) {
        guard let call = self.call,
                sessionId == call.sessionID,
              let provider = provider else {
            return
        }
        
        provider.reportCall(with: call.uuid, endedAt: Date(), reason: reason)
        closeCall(call.sessionID)
    }
    
    func reportAcceptCall(sessionId: String) {
        guard let provider = self.provider,
              let call = self.call else {
            return
        }
        
        let actions = provider.pendingCallActions(of: CXAnswerCallAction.self, withCall: call.uuid)
        for action in actions {
            if let answer = action as? CXAnswerCallAction {
                answer.fulfill(withDateConnected: Date())
                break
            }
        }
    }
    
    func reportUpdateCall(sessionId: String, title: String?) {
        guard let title = title,
              let provider = provider,
              let callUUID = UUID(uuidString: sessionId) else {
            return
        }
        
        let update = CXCallUpdate()
        update.remoteHandle = handle(withText:title)
        update.localizedCallerName = title
        provider.reportCall(with: callUUID, updated: update)
    }
    
    func muteAudio(_ mute: Bool, forCall sessionId: String) {
        guard let call = call, call.sessionID == sessionId else {
            return
        }
        
        let  action = CXSetMutedCallAction(call: call.uuid, muted: mute)
        let transaction = CXTransaction(action: action)
        callController.request(transaction) { error in
            guard let error = error else {
                return
            }
            debugPrint("\(#function) Error: \(error)")
        }
    }
    
    // MARK: - Helpers
    private func handle(withText text: String) -> CXHandle {
        return CXHandle(type: .generic, value: text)
    }
    
    private func callUpdate(withTitle title: String, hasVideo: Bool) -> CXCallUpdate {
        let update = CXCallUpdate()
        update.localizedCallerName = title
        update.supportsHolding = false
        update.supportsGrouping = false
        update.supportsUngrouping = false
        update.supportsDTMF = false
        update.hasVideo = hasVideo
        return update
    }
    
    private func closeCall(_ sessionID: String) {
        qbAudioSession.isAudioEnabled = false
        qbAudioSession.useManualAudio = false
        
        if reportEndCall {
            delegate?.callKit(self, didEndCall: sessionID)
        } else {
            delegate?.callKit(self, didTapRedject: sessionID)
        }
        
        reportEndCall = false
        self.call = nil
    }
    
    private func updateAudioSessionConfiguration(_ hasVideo: Bool) {
        let configuration = QBRTCAudioSessionConfiguration()
        configuration.categoryOptions.insert(.duckOthers)

        // adding blutetooth support
        configuration.categoryOptions.insert(.allowBluetooth)
        configuration.categoryOptions.insert(.allowBluetoothA2DP)

        // adding airplay support
        configuration.categoryOptions.insert(.allowAirPlay)
        
        if hasVideo == true {
            // setting mode to video chat to enable airplay audio and speaker only
            configuration.mode = AVAudioSession.Mode.videoChat.rawValue
        }
        qbAudioSession.setConfiguration(configuration)
    }
}

// MARK: - CXProviderDelegate
extension CallKitManager: CXProviderDelegate {
    
    func providerDidReset(_ provider: CXProvider) {
        call = nil
    }
    
    func provider(_ provider: CXProvider, perform action: CXStartCallAction) {
        guard let call = call,
              action.callUUID == call.uuid else {
            action.fail()
            return
        }
        updateAudioSessionConfiguration(call.hasVideo)
        
        action.fulfill()
    }
    
    func provider(_ provider: CXProvider, perform action: CXAnswerCallAction) {
        guard let call = call,
              action.callUUID == call.uuid else {
            action.fail()
            return
        }
        updateAudioSessionConfiguration(call.hasVideo)
        
        delegate?.callKit(self, didTapAnswer: call.sessionID)
    }
    
    func provider(_ provider: CXProvider, perform action: CXEndCallAction) {
        guard let call = call,
              action.callUUID == call.uuid else {
            action.fail()
            return
        }
        closeCall(call.sessionID)
        
        action.fulfill()
    }
    
    func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession) {
        qbAudioSession.audioSessionDidActivate(audioSession)
        qbAudioSession.isAudioEnabled = true
    }
    
    func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession) {
        qbAudioSession.audioSessionDidDeactivate(audioSession)
    }
    
    func provider(_ provider: CXProvider, perform action: CXSetMutedCallAction) {
        guard let call = call,
              action.callUUID == call.uuid else {
            action.fail()
            return
        }
        actionDelegate?.callKit(self, didTapMute: action.isMuted)
        action.fulfill()
    }
}
